x/protorev/keeper/hooks.go

```diff
@@ -0,0 +1,235 @@
+	h.k.afterPoolCreated(ctx, poolId)
+}
+
+// AfterInitialPoolPositionCreated is a noop.
+func (h Hooks) AfterInitialPoolPositionCreated(ctx sdk.Context, sender sdk.AccAddress, poolId uint64) {
+}
+
+// AfterLastPoolPositionRemoved is a noop.
+func (h Hooks) AfterLastPoolPositionRemoved(ctx sdk.Context, sender sdk.AccAddress, poolId uint64) {
+}
+
+// AfterConcentratedPoolSwap stores swaps to be checked by protorev given the coins swapped in the pool.
+func (h Hooks) AfterConcentratedPoolSwap(ctx sdk.Context, sender sdk.AccAddress, poolId uint64, input sdk.Coins, output sdk.Coins) {
+	// Checked to avoid future unintended behavior based on how the hook is called
+	if len(input) != 1 || len(output) != 1 {
+		return
+	}
+
+	h.k.storeSwap(ctx, poolId, input[0].Denom, output[0].Denom)
+}
+
+// ----------------------------------------------------------------------------
+// HELPER METHODS
+// ----------------------------------------------------------------------------
+
+// afterPoolCreated checks if the new pool should be stored as the highest liquidity pool
+// for any of the base denoms, and stores it if so.
+func (k Keeper) afterPoolCreated(ctx sdk.Context, poolId uint64) {
+	baseDenoms, err := k.GetAllBaseDenoms(ctx)
+	if err != nil {
+		ctx.Logger().Error("Protorev error getting base denoms in AfterCFMMPoolCreated hook", err)
+		return
+	}
+
+	baseDenomMap := make(map[string]bool)
+	for _, baseDenom := range baseDenoms {
+		baseDenomMap[baseDenom.Denom] = true
+	}
+
+	pool, err := k.poolmanagerKeeper.GetPool(ctx, poolId)
+	if err != nil {
+		ctx.Logger().Error("Protorev error getting pool in AfterCFMMPoolCreated hook", err)
+		return
+	}
+
+	coins := k.getCoinsFromPool(ctx, pool, poolId)
+
+	// Pool must be active and the number of coins must be 2
+	if pool.IsActive(ctx) && len(coins) == 2 {
+		tokenA := coins[0]
+		tokenB := coins[1]
+
+		if baseDenomMap[tokenA.Denom] {
+			k.compareAndStorePool(ctx, poolId, tokenA, tokenB)
+		}
+		if baseDenomMap[tokenB.Denom] {
+			k.compareAndStorePool(ctx, poolId, tokenB, tokenA)
+		}
+	}
+}
+
+// getCoinsFromPool gets the coins from the pool, handling concentrated pools differently since they can have 0 liquidity.
+func (k Keeper) getCoinsFromPool(ctx sdk.Context, pool poolmanagertypes.PoolI, poolId uint64) sdk.Coins {
+	coins, err := k.poolmanagerKeeper.GetTotalPoolLiquidity(ctx, poolId)
+	if err != nil {
+		ctx.Logger().Error("Protorev error getting pool liquidity in AfterCFMMPoolCreated hook", err)
+		return sdk.Coins{}
+	}
+
+	if coins == nil {
+		if pool.GetType() == poolmanagertypes.Concentrated {
+			clPool, ok := pool.(cltypes.ConcentratedPoolExtension)
+			if !ok {
+				ctx.Logger().Error("Protorev error casting pool to concentrated pool in AfterCFMMPoolCreated hook")
+				return sdk.Coins{}
+			}
+			coins = sdk.Coins{sdk.NewCoin(clPool.GetToken0(), sdk.NewInt(0)), sdk.NewCoin(clPool.GetToken1(), sdk.NewInt(0))}
```

Comment:
So in hindsight this is actually purposeful, since NewCoins() sanitizes coins with a 0 value, so ends up returning a blank coins object.

Reverted here: https://github.com/osmosis-labs/osmosis/pull/5045/commits/6faeab31fe71d17a5963497542c0532f174db2fb

This change what caused the test to fail

Reply:
i see gotchu

Reply:
I'm worried about bypassing this constructor validation of zero coins. This might lead to panic elsewhere as the system evolves.

Can we instead special handle CL pools in the caller?
- if CL pool and is active => tokenA := clPool.GetToken0()  tokenB := clPool.GetToken1()

Reply:
My take on resolving: https://github.com/osmosis-labs/osmosis/pull/5045#issuecomment-1572845639